### PR TITLE
rpcd-mod-poe: add package

### PR DESCRIPTION
**Pull request by aparcar**
Maintainer: me

simple package to manipulate gpio pins via the web interface. This was
tested on a UBNT EdgePing R6 turining on and off the 4 PoE ports
(496-500).

A better solution would be to have a mapping like the following

    eth0 -> 496
    eth1 -> 497
    eth2 -> 498
    eth3 -> 499
    eth4 -> 500

in `target/linux/ramips/dts/UBNT-ERX-SFP.dts` and then use a command like 

    ubus call poe off { "port": "eth1" }

Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

File: utils/rpcd-mod-poe/files/poe.acl [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+	"poe": {
+		"description": "manipulate gpio pins to enable/disable poe",
+		"read": {
+			"ubus": {
+				"poe": [
+					"on",
```
Comment (jow-) [inline utils/rpcd-mod-poe/files/poe.acl]: The `on` and `off` procedures have side effects and should not be part of the read scope. Declare a `write` scope and put `on` & `off` there while only keeping `status` in `"read": { ... }`

File: utils/rpcd-mod-poe/Makefile [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+PKG_LICENSE:=GPL-2.0
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/rpcd-mod-poe
+  SECTION:=utils
+  CATEGORY:=Base system
+  TITLE:=ubus poe interface
+  MAINTAINER:=Paul Spooren <mail@aparcar.org>
+  DEPENDS:=rpcd kmod-i2c-gpio-custom
```
Comment (jow-) [inline utils/rpcd-mod-poe/Makefile]: Please use selecting dependencies `DEPENDS:=+rpcd +kmod-i2c-gpio-custom`

File: utils/rpcd-mod-poe/files/poe.rpcd [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	json_add_object "off"
+	json_add_int "gpio"
+	json_close_object
+	json_add_object "status"
+	json_add_int "gpio"
+	json_close_object
+	json_dump
+	;;
+call)
+	read input;
```
Comment (jow-) [inline utils/rpcd-mod-poe/files/poe.rpcd]: Please use a consistent style here (either end all commands on a semicolon or none)

File: utils/rpcd-mod-poe/Makefile (hunk 2 above)
Comment (jow-) [reply utils/rpcd-mod-poe/Makefile]: I suggest to captialize the title and use proper casing for PoE as well, e.g. `TITLE:=OpenWrt ubus PoE interface`

File: utils/rpcd-mod-poe/Makefile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+endef
+
+define Package/rpcd-mod-poe/description
+	add ubus call to manipulate gpio pins that control poe
```
Comment (jow-) [inline utils/rpcd-mod-poe/Makefile]: I'd reword that into something like:

```
Expose ubus procedures to query, enable and disable the per-port PoE power on
Ubiquity devices and similar platforms.
```

**Comment by aparcar** [comment]
@jow- as always, thank you very much for the review!

Could you give an opinion about the mapping I mentioned in the PR message? That would greatly improve the usage, as I could extend the luci switch view with some buttons for on/off/powercycling...

**Comment by jow-** [comment]
I have no opinion on the mapping, you should ask the paltform maintainers on their opinion, upstream-ability of such changes etc.

**Comment by aparcar** [comment]
@jow- if it's fine with you pleas accept this PR or mark your changes as resolved. I'll ask @dangowrt for further comments

**Comment by aparcar** [comment]
Reworked to modify UCI vales rather than raw GPIO

the `list` output looks like this, to be used later in luci

```
{
   "poe_power_port0":{
      "name":"PoE Power Port0",
      "gpio_pin":496,
      "active":1
   },
   "poe_power_port1":{
      "name":"PoE Power Port1",
      "gpio_pin":497,
      "active":0
   },
   "poe_power_port2":{
      "name":"PoE Power Port2",
      "gpio_pin":498,
      "active":0
   },
   "poe_power_port3":{
      "name":"PoE Power Port3",
      "gpio_pin":499,
      "active":0
   },
   "poe_power_port4":{
      "name":"PoE Power Port4",
      "gpio_pin":500,
      "active":1
   }
}
```

**Comment by aparcar** [comment]
@dangowrt please have a look

**Comment by dangowrt** [comment]
You kinda moved two steps backwards since we last spoke. Again, I don't believe they value of gpio_pin should be used in configuration or even front-end -- it's a dynamically allocated number for this gpio-chip and not persistent by design. And it's ugly, require old-school uci-defaults to handle board-specific details ie. knowledge about that specific board will then need to be included for all boards on a platform instead of being shipped in DTS which is much more clean.
See also https://github.com/openwrt/openwrt/pull/1814

**Comment by aparcar** [comment]
I'd be in favor to merge this request for now, I'll create a Luci app and once it's fixed on the lower levels, I'll rework this package. @dangowrt @jow-

**Comment by aparcar** [comment]
ping?

**Comment by dangowrt** [comment]
@aparcar merge yourself if it's ready to go

**Comment by rmilecki** [comment]
Why would we ever want to expose GPIO number? I don't think user should bother about that at all.
If I want to enable/disable a particular LED, I use `/sys/class/leds/` and I don't care what's the low-level implementation.
What if some device has PoE ports controlled another way?
I'd really prefer to see some well-designed & generic solution for that.

File: utils/rpcd-mod-poe/files/poe.rpcd [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		json_init
+		json_add_string "error" "gpio $gpio not found"
+		json_dump
+	}
+
+	case "$2" in
+	on)
+		uci set system.poe_power_port${port}.value=1
+		uci commit system
+		/etc/init.d/gpio_switch restart
```
Comment (karlp) [inline utils/rpcd-mod-poe/files/poe.rpcd]: what package is gpio_switch in? it's not in the deps?

File: utils/rpcd-mod-poe/Makefile [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  SECTION:=utils
+  CATEGORY:=Base system
+  TITLE:=OpenWrt ubus PoE interface
+  MAINTAINER:=Paul Spooren <mail@aparcar.org>
+  DEPENDS:=+rpcd +kmod-i2c-gpio-custom
+endef
+
+define Package/rpcd-mod-poe/description
+	Expose ubus procedures to query, enable and disable the per-port PoE power
+	on Ubiquity devices and similar platforms.
```
Comment (karlp) [inline utils/rpcd-mod-poe/Makefile]: the name seems both overly specific when it's just doign gpio toggling, and overly generic, when it's only handling PoE for thigns using this particular gpio_switch config for ubiquity.  DO you really just want a generic gpio control?

Comment (aparcar) [reply utils/rpcd-mod-poe/Makefile]: I want to power cycle PoE ports, nothing more. Please help me find a clean solution for that. I'm happy to use device tree or whatever else.

Comment (dangowrt) [reply utils/rpcd-mod-poe/Makefile]: For real PoE there are vendor-specific hacks, see
https://docs.cumulusnetworks.com/display/DOCS/Power+over+Ethernet+-+PoE
for inspiration.

Maybe the information about the mapping of PoE-GPIOs to Ethernet ports can somehow be described in device-tree in a way the kernel wouldn't care but make it possible for userspace to retrieve that information? That would already make it possible to describe eg. the PoE-passthrough of simple ubnt devices.
I reckon for a clean solution for ER-X we will have to wait for DSA finding use for MT7621 -- the MT7630 driver was recently extended to [support the MT7621](https://github.com/torvalds/linux/commit/ddda1ac116c852bb969541ed53cffef7255c4961#diff-5aed940c7b18a8002cacb9dce5bd5c1f) upstream, so all we will have to do is to convert the existing swconfig LAN/WAN setup into DSA for all boards...

Comment (aparcar) [reply utils/rpcd-mod-poe/Makefile]: How much work is it? More like ~2 month or 2 years?
